tests/test_perf.py

```diff
@@ -49,3 +55,106 @@ def test_perf(
+    application_docker_container.restart(timeout=0)
+    application_docker_container.reload()  # post restart
+    return cast(int, application_docker_container.attrs["State"]["Pid"])
+
+
+def _assert_comm_in_profile(profiler: SystemProfiler, application_pid: int, exec_comm: bool) -> None:
+    collapsed = snapshot_pid_collapsed(profiler, application_pid)
+    # native is the original comm
+    # oative is the changed comm of the main thread
+    # pative is the changed comm of other threads
+    if exec_comm:
+        assert is_function_in_collapsed("native;", collapsed)
+        assert not is_function_in_collapsed("oative;", collapsed)
+        assert not is_function_in_collapsed("pative;", collapsed)
+    else:
+        assert not is_function_in_collapsed("native;", collapsed)
+        assert is_function_in_collapsed("oative;", collapsed)
+        assert not is_function_in_collapsed("pative;", collapsed)
+
+
+@pytest.mark.parametrize("runtime", ["native_change_comm"])
+@pytest.mark.parametrize("perf_mode", ["fp"])  # only fp is enough
+@pytest.mark.parametrize("in_container", [True])  # native app is built only for container
+def test_perf_comm_change(
+    system_profiler: SystemProfiler,
+    application_pid: int,
+    application_docker_container: Container,
+    perf_mode: str,
+) -> None:
+    """
+    Runs a program that changes its comm (i.e comm of the main thread).
+    The stack output by perf should use the original name, not the changed one.
+
+    This tests our modification to perf here:
+    https://github.com/Granulate/linux/commit/40a7823cf90a7e69ce8af88d224dfdd7e371de2d
+
+    perf records "comm" events - changes of task comms. Samples printed for a thread will use the current comm
+    perf understands that it has. Our modification makes it select a PERF_RECORD_COMM event that has the "exec"
+    flag. Below are 2 events from the native program - the first is the name it was executed with, then
+    the rename. If perf has started before the app, we expect to see the exec name and not the current name.
+
+        native 1901079 [010] 984807.648525: PERF_RECORD_COMM exec: native:1901079/1901079
+        oative 1901079 [010] 984807.648792: PERF_RECORD_COMM: oative:1901079/1901079
+
+    If perf started after the app, we will see the changed name (although I'd prefer to see the original name, but
+    I'm not sure it can be done, i.e is this info even kept anywhere).
+    """
+    with system_profiler as profiler:
+        # first run - we get the changed name, because the app started before perf began recording.
+        _assert_comm_in_profile(profiler, application_pid, False)
+
+        # now, while perf is still running & recording, we restart the app
+        application_pid = _restart_app_container(application_docker_container)
+
+        # second run - we get the original name, because the app started after perf began recording.
+        _assert_comm_in_profile(profiler, application_pid, True)
```

Comment:
Nope. `snapshot` is - run gProfiler for a single session. Now for `perf` and `PyPerf` and `phpspy`, since they're continuous, if you:
```
snapshot()
xxx()
snapshot()
```
then it's racy because after `snapshot()` returns, they continue collecting samples until `xxx()` happens.

BUT here, we use `snapshot_pid_collapsed` to get the samples from a single PID, and we actually restart the app and switch to use the new PID. So nope, not racy

Reply:
Because all the invocations are under the same `with` - it appeared confusing. thx.